include/cuda/std/detail/libcxx/include/__concepts/constructible.h

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___CONCEPTS_CONSTRUCTIBLE_H
+#define _LIBCUDACXX___CONCEPTS_CONSTRUCTIBLE_H
+
+#ifndef __cuda_std__
+#include <__concepts/convertible_to.h>
+#include <__concepts/destructible.h>
+#include <__config>
+#include <type_traits>
+#else
+#include "convertible_to.h"
+#include "destructible.h"
+#endif //__cuda_std__
+
+#if !defined(_LIBCUDACXX_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+#if _LIBCUDACXX_STD_VER > 17
+
+// [concept.constructible]
+template<class _Tp, class... _Args>
+concept constructible_from =
+    destructible<_Tp> && is_constructible_v<_Tp, _Args...>;
+
+// [concept.default.init]
+template<class _Tp>
+concept __default_initializable = requires { ::new _Tp; };
+
+template<class _Tp>
+concept default_initializable = constructible_from<_Tp> &&
+    requires { _Tp{}; } && __default_initializable<_Tp>;
+
+// [concept.moveconstructible]
+template<class _Tp>
+concept move_constructible =
+  constructible_from<_Tp, _Tp> && convertible_to<_Tp, _Tp>;
+
+// [concept.copyconstructible]
+template<class _Tp>
+concept copy_constructible =
+  move_constructible<_Tp> &&
+  constructible_from<_Tp, _Tp&> && convertible_to<_Tp&, _Tp> &&
+  constructible_from<_Tp, const _Tp&> && convertible_to<const _Tp&, _Tp> &&
+  constructible_from<_Tp, const _Tp> && convertible_to<const _Tp, _Tp>;
+
+#else if _LIBCUDACXX_STD_VER > 11
+
+template<class _Tp, class... _Args>
+_LIBCUDACXX_CONCEPT constructible_from =
+    destructible<_Tp> && is_constructible_v<_Tp, _Args...>;
+
+template<class _Tp>
+_LIBCUDACXX_CONCEPT_FRAGMENT(
+  __default_initializable_,
+  requires() //
+  ((::new _Tp)));
+  
+template<class _Tp>
+_LIBCUDACXX_CONCEPT __default_initializable = _LIBCUDACXX_FRAGMENT(__default_initializable_, _Tp);
+  
+template<class _Tp>
+_LIBCUDACXX_CONCEPT_FRAGMENT(
+  _Default_initializable_,
+  requires() //
+  (_Tp{}));
+template<class _Tp>
+_LIBCUDACXX_CONCEPT default_initializable = constructible_from<_Tp> &&
+    _LIBCUDACXX_FRAGMENT(_Default_initializable_, _Tp) && __default_initializable<_Tp>;
+
+// [concept.moveconstructible]
+template<class _Tp>
+_LIBCUDACXX_CONCEPT move_constructible =
+  constructible_from<_Tp, _Tp> && convertible_to<_Tp, _Tp>;
+
+// [concept.copyconstructible]
+template<class _Tp>
+_LIBCUDACXX_CONCEPT copy_constructible =
+  move_constructible<_Tp> &&
+  constructible_from<_Tp, add_lvalue_reference_t<_Tp>> && convertible_to<add_lvalue_reference_t<_Tp>, _Tp> &&
```

Comment:
Aha! Yeah that makes sense. And I assume the cases of `add_pointer` are because the types may be unstripped references so we also can't directly add a pointer, unlike in the 20 mode.